### ISSUE TITLE
Maximum call stack size exceeded when pasting large content into ace in 64-bit chrome

### ISSUE THREAD

**Issue by glajchs** [ajaxorg/ace#2374]
I was having this problem primarily when pasting very large JSON documents (130k lines) and running ace in JSON syntax mode, however I was able to reproduce it with other language types as well (javascript, though it required some work).

There's already protection against this problem, however it seems to have regressed for 64 bit browsers (or at least for chrome).  Basically the native Array.push.apply() function call causes this "Worker RangeError: Maximum call stack size exceeded".  Here's the line number in source (though it will actually show up in the worker-json.js file:
https://github.com/ajaxorg/ace/blob/master/lib/ace/document.js#L323

The protection against this is just a few lines up at:
https://github.com/ajaxorg/ace/blob/master/lib/ace/document.js#L316
Three of the lines reference a hard limit of the maximum number of lines to pass into the array below, of the number 0xF000 (or 61440 in decimal).  Presumably this was approaching the limit of 0xFFFF (65535), and to give some buffer to allow for however many stack frames already existed the value was moved down to 0xF000.  I've found that (in my case) somewhere between 31000 and 32000 the pasting works again.  I've found this only in the 64 bit version of chrome.  I found this issue on Linux Version 41.0.2272.64 beta (64-bit), and I had a friend test on Mac with chrome v40 64 bit and he saw it too.  I was unable to reproduce in a Windows VM with chrome v40 32-bit (both 32 bit OS and browser).  I have not tested other browsers yet.

I'm not exactly sure why this is, it does seem a bit counter-intuitive that the larger memory space would have a lower stack size, but I'd guess that in 64 bit, some memory references are now twice as large as they were in 32 bit, which allows for fewer stack frames before hitting an overall size limit.

Instead of just patching this to be just below the 32k value, I think making it something much lower like 1k would be better.  After all this would chop up the stack frame to be much smaller overall, while still keeping the number of iterations through your loop to a pretty low number.

TL;DR:
Locally I modified my worker-json.js to make the 3 instances of 0xF000 to be either 31000 (in decimal), or 1000 (in decimal), and I no longer had this issue.

**Comment by nightwing** [comment]
Setting it to 1000 makes insert significantly slower. I've increased it to 20000 for now. 
If browsers reduce maximum call stack further, we'll have to stop using splice, and use a tree instead of a simple array.
